src/Algebra/Solver/IdempotentCommutativeMonoid.agda

```diff
@@ -10,208 +10,35 @@
 
 open import Algebra.Bundles using (IdempotentCommutativeMonoid)
 
-open import Data.Bool as Bool using (Bool; true; false; if_then_else_; _∨_)
-open import Data.Fin.Base using (Fin; zero; suc)
-open import Data.Maybe.Base as Maybe
-  using (Maybe; From-just; from-just)
-open import Data.Nat.Base as ℕ using (ℕ; zero; suc; _+_)
-open import Data.Product.Base using (_×_; uncurry)
-open import Data.Vec.Base using (Vec; []; _∷_; lookup; replicate)
-
-open import Function.Base using (_∘_)
-
-import Relation.Binary.Reasoning.Setoid  as ≈-Reasoning
-import Relation.Binary.Reflection            as Reflection
-import Data.Vec.Relation.Binary.Pointwise.Inductive as Pointwise
-
-open import Relation.Binary.Consequences using (dec⇒weaklyDec)
-open import Relation.Binary.PropositionalEquality.Core as ≡ using (_≡_)
-open import Relation.Binary.PropositionalEquality.Properties using (decSetoid)
-open import Relation.Nullary.Decidable as Dec using (Dec)
-
-
 module Algebra.Solver.IdempotentCommutativeMonoid
-  {m₁ m₂} (M : IdempotentCommutativeMonoid m₁ m₂) where
-
-open IdempotentCommutativeMonoid M
-open ≈-Reasoning setoid
-
-private
-  variable
-    n : ℕ
-
-------------------------------------------------------------------------
--- Monoid expressions
-
--- There is one constructor for every operation, plus one for
--- variables; there may be at most n variables.
-
-infixr 5 _⊕_
-infixr 10 _•_
-
-data Expr (n : ℕ) : Set where
-  var : Fin n → Expr n
-  id  : Expr n
-  _⊕_ : Expr n → Expr n → Expr n
+  {c ℓ} (M : IdempotentCommutativeMonoid c ℓ) where
 
--- An environment contains one value for every variable.
+import Algebra.Solver.IdempotentCommutativeMonoid.Normal as Normal
+import Algebra.Solver.Monoid.Tactic as Tactic
 
-Env : ℕ → Set _
-Env n = Vec Carrier n
-
--- The semantics of an expression is a function from an environment to
--- a value.
-
-⟦_⟧ : ∀ {n} → Expr n → Env n → Carrier
-⟦ var x   ⟧ ρ = lookup ρ x
-⟦ id      ⟧ ρ = ε
-⟦ e₁ ⊕ e₂ ⟧ ρ = ⟦ e₁ ⟧ ρ ∙ ⟦ e₂ ⟧ ρ
+open IdempotentCommutativeMonoid M using (monoid)
 
 ------------------------------------------------------------------------
 -- Normal forms
 
--- A normal form is a vector of bits (a set).
-
-Normal : ℕ → Set
-Normal n = Vec Bool n
-
--- The semantics of a normal form.
-
-⟦_⟧⇓ : Normal n → Env n → Carrier
-⟦ []    ⟧⇓ _ = ε
-⟦ b ∷ v ⟧⇓ (a ∷ ρ) = if b then a ∙ (⟦ v ⟧⇓ ρ) else (⟦ v ⟧⇓ ρ)
-
-------------------------------------------------------------------------
--- Constructions on normal forms
-
--- The empty set.
-
-empty : Normal n
-empty = replicate _ false
-
--- A singleton set.
-
-sg : (i : Fin n) → Normal n
-sg zero    = true ∷ empty
-sg (suc i) = false ∷ sg i
-
--- The composition of normal forms.
-
-_•_ : (v w : Normal n) → Normal n
-[]      • []      = []
-(l ∷ v) • (m ∷ w) = (l ∨ m) ∷ v • w
+open module N = Normal M public
```

Comment:
Again is there a reason we are deliberately exporting a module named `N`? (See other files as well)

Reply:
Fixed (I hope!)

Reply:
Fixed